provider/azure/environ.go

```diff
@@ -1074,7 +1074,56 @@ func (env *azureEnviron) instances(
 
 // AdoptResources is part of the Environ interface.
 func (env *azureEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
-	return errors.NotImplementedf("AdoptResources")
+	groupClient := resources.GroupsClient{env.resources}
+	resourceClient := resources.Client{env.resources}
+	var failed []string
+
+	resources, err := groupClient.ListResources(env.resourceGroup, "", nil)
+	if err != nil {
+		return errors.Annotate(err, "listing resources")
+	}
+	for resources.Value != nil {
+		for _, resource := range *resources.Value {
+
+			providerNamespace, parentPath, err := splitResourceType(*resource.Type)
+			if err != nil {
+				failed = append(failed, *resource.Name)
+				logger.Errorf("error getting namespace and path for type %q: %v", *resource.Type, err)
+			}
+
+			(*resource.Tags)[tags.JujuController] = &controllerUUID
```

Comment:
What I meant was: if this fails half way through, e.g. due to a network failure, will/can it be called again? If so, we should handle a half-processed resource group.